core/src/main/scala/org/apache/spark/storage/PushBasedFetchHelper.scala

```diff
@@ -0,0 +1,336 @@
+    val cachedMergerDirs = hostLocalDirManager.getCachedHostLocalDirs.get(
+      SHUFFLE_MERGER_IDENTIFIER)
+    if (cachedMergerDirs.isDefined) {
+      logDebug(s"Fetching local merged blocks with cached executors dir: " +
+        s"${cachedMergerDirs.get.mkString(", ")}")
+      mergedLocalBlocks.foreach(blockId =>
+        fetchMergedLocalBlock(blockId, cachedMergerDirs.get, localShuffleMergerBlockMgrId))
+    } else {
+      logDebug(s"Asynchronous fetching local merged blocks without cached executors dir")
+      hostLocalDirManager.getHostLocalDirs(localShuffleMergerBlockMgrId.host,
+        localShuffleMergerBlockMgrId.port, Array(SHUFFLE_MERGER_IDENTIFIER)) {
+        case Success(dirs) =>
+          mergedLocalBlocks.takeWhile {
+            blockId =>
+              logDebug(s"Successfully fetched local dirs: " +
+                s"${dirs.get(SHUFFLE_MERGER_IDENTIFIER).mkString(", ")}")
+              fetchMergedLocalBlock(blockId, dirs(SHUFFLE_MERGER_IDENTIFIER),
+                localShuffleMergerBlockMgrId)
+          }
+          logDebug(s"Got local merged blocks (without cached executors' dir) in " +
+            s"${TimeUnit.NANOSECONDS.toMillis(System.nanoTime() - startTimeNs)} ms")
+        case Failure(throwable) =>
+          // If we see an exception with getting the local dirs for local merged blocks,
+          // we fallback to fetch the original unmerged blocks. We do not report block fetch
+          // failure.
+          logWarning(s"Error occurred while getting the local dirs for local merged " +
+            s"blocks: ${mergedLocalBlocks.mkString(", ")}. Fetch the original blocks instead",
+            throwable)
+          mergedLocalBlocks.foreach(
+            blockId => iterator.addToResultsQueue(FallbackOnMergedFailureFetchResult(
+              blockId, localShuffleMergerBlockMgrId, 0, isNetworkReqDone = false))
+          )
+      }
+    }
+  }
+
+  /**
+   * Fetch a single local merged block generated. This can also be executed by the task thread as
+   * well as the netty thread.
+   * @param blockId ShuffleBlockId to be fetched
+   * @param localDirs Local directories where the merged shuffle files are stored
+   * @param blockManagerId BlockManagerId
+   * @return Boolean represents successful or failed fetch
+   */
+  private[this] def fetchMergedLocalBlock(
+      blockId: BlockId,
+      localDirs: Array[String],
+      blockManagerId: BlockManagerId): Boolean = {
+    try {
+      val shuffleBlockId = blockId.asInstanceOf[ShuffleBlockId]
+      val chunksMeta = blockManager.getLocalMergedBlockMeta(shuffleBlockId, localDirs)
+        .readChunkBitmaps()
+      // Fetch local merged shuffle block data as multiple chunks
+      val bufs: Seq[ManagedBuffer] = blockManager.getLocalMergedBlockData(shuffleBlockId, localDirs)
+      // Update total number of blocks to fetch, reflecting the multiple local chunks
+      iterator.incrementNumBlocksToFetch(bufs.size - 1)
```

Comment:
This is because for each local merged block, `numBlocksToFetch` is already incremented by 1 in `partitionBlocksByFetchMode`.

Reply:
Since this can get called by netty thread as well, `numBlocksToFetch` need to be protected. I will fix this. I think it's better that only task thread modifies this so probably push an existing or another result to the results queue would be the best approach.

Reply:
> This is because for each local merged block, numBlocksToFetch is already incremented by 1 in partitionBlocksByFetchMode.

Oh...that's easy to miss.  I think it's better to update them separately for readability. For example,

```scala
val chunksMeta = blockManager.getLocalMergedBlockMeta(shuffleBlockId, localDirs)
     .readChunkBitmaps()
// for each local merged block, numBlocksToFetch is already incremented by 1 in partitionBlocksByFetchMode.
// So we should decrease it after getting the target metdata.
iterator.updateNumBlocksToFetch(-1)
// Fetch local merged shuffle block data as multiple chunks
val bufs: Seq[ManagedBuffer] = blockManager.getLocalMergedBlockData(shuffleBlockId, localDirs)
// Update total number of blocks to fetch, reflecting the multiple local chunks
iterator.updateNumBlocksToFetch(bufs.size)
```

Reply:
I was confused as well, until I read the comment explaining why `size - 1` (and I had to validate this was indeed valid while reviewing) :-)
Explicitly decrementing and then incrementing by size will make it clear given the cost is negligible.
I like the proposal @Ngone51.

Reply:
I have made this change but now this happens when the iterator processes ` PushMergedLocalMetaFetchResult`

Reply:
Resolving this. All the updates to `numBlocksToFetch` as well as `chunksMetaMap` are done by task thread now.  So also reverted the change that made `chunksMetaMap` concurrent.